### PR TITLE
feat: use functional induction support to avoid long compile times

### PR DESCRIPTION
**Pull request by tobiasgrosser**

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
This patch is WIP and currently stuck at:

```lean4
-- motive : ∀ (ie : Expr d Γ_out EffectKind.pure (DialectSignature.outTy matchExpr.op))
-- (hs : ∃ (h : ie.op = matchExpr.op), ie.regArgs = ⋯ ▸ matchExpr.regArgs) (varMap : Mapping Δ_in Γ_out),
-- matchArg lets matchLets ie.args (⋯ ▸ matchExpr.args) ma = some varMap → ma.entries ⊆ varMap.entries
apply motive
· -- goal:
  -- matchArg lets matchLets ?pos.intro.ie✝.args (⋯ ▸ matchExpr.args) ma = some varMap
  sorry
· -- goal:
  -- Expr d Γ_out EffectKind.pure (DialectSignature.outTy matchExpr.op)
  --
  -- does the 'aa' hypothesis help? I am not sure how to get rid of the 'Option'?
  --   aa:  Option (Expr d Γ_out EffectKind.pure (DialectSignature.outTy matchExpr.op))
  sorry
· -- goal
  -- ∃ (h : (sorryAx (Expr d Γ_out EffectKind.pure (DialectSignature.outTy matchExpr.op))).op = matchExpr.op),
  -- (sorryAx (Expr d Γ_out EffectKind.pure (DialectSignature.outTy matchExpr.op))).regArgs = ⋯ ▸ matchExpr.regArgs
  sorry
```

help is much appreciated.

**Comment by alexkeizer** [comment]
@tobiasgrosser I pushed this through, and replaced the old mutual proofs with the new functional induction one

File: SSA/Core/Framework.lean [hunk 1]
```diff
@@ -1714,6 +1714,9 @@ end TermModel
 
 -/
 
+
+
+set_option maxHeartbeats 9999999 in
```
Comment (tobiasgrosser) [inline SSA/Core/Framework.lean]: Do we still need this?

**Comment by tobiasgrosser** [comment]
@alexkeizer, this is lovely. Thank you.

**Comment by tobiasgrosser** [comment]
This is good to merge.

**Pull request merged**